### PR TITLE
Doc: multithreading: explicit why GDALDataset is not thread-safe

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

File: doc/source/user/multithreading.rst [hunk 1]
```diff
@@ -31,13 +31,22 @@ on different :cpp:class:`GDALRasterBand` instances owned by the same
...
 distinct GDALDataset). Similarly for a GDALDataset owning several :cpp:class:`OGRLayer`.
 
+The reason is that most implementations of GDALDataset or GDALRasterBand
+are not state-less. A GDALDataset typically owns a file handle,
+and performs seek/read operations on it, thus not allowing concurrent access.
+Block cache related structures for a given GDALDataset are not thread-safe.
+Drivers also often implement lazy initialization strategies to access various
+metadata which are resolved only the first time the method to access them is
+invoked. Drivers may also rely on third-party libraries that expose objects
+that are not tread-safe.
```
Comment (Ryanf55) [inline doc/source/user/multithreading.rst:41]:
```suggestion
that are not thread-safe.
```

File: doc/source/user/multithreading.rst [hunk 2]
```diff
@@ -31,13 +31,22 @@ on different :cpp:class:`GDALRasterBand` instances owned by the same
 :cpp:class:`GDALDataset` instance (each thread should instead manipulate a
 distinct GDALDataset). Similarly for a GDALDataset owning several :cpp:class:`OGRLayer`.
 
+The reason is that most implementations of GDALDataset or GDALRasterBand
+are not state-less. A GDALDataset typically owns a file handle,
```
Comment (sgillies) [inline doc/source/user/multithreading.rst:35]: How about the following, removing the double negative?

```suggestion
are stateful. A GDALDataset typically owns a file handle,
```

File: doc/source/user/multithreading.rst [hunk 3]
```diff
@@ -31,13 +31,22 @@ on different :cpp:class:`GDALRasterBand` instances owned by the same
...
+
 Those restrictions apply to the C and C++ ABI, and all languages bindings (unless
 they would take special precautions to serialize calls)
 
 GDAL block cache and multi-threading
 ------------------------------------
 
-The current design of the GDAL raster block cache make it appropriate to
+The current design of the GDAL raster block cache makes it appropriate to
 read several datasets from several threads. However performance issues may
```
Comment (sgillies) [inline doc/source/user/multithreading.rst:50]:
```suggestion
The current design of the GDAL raster block cache allows concurrent reads of several datasets. However performance issues may
```

**Review by sgillies — APPROVED** [review]
Thank you for this @rouault ! I gave a couple minor suggestions about wording.

**Pull request merged**